<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Configuration &mdash; MINORg 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial" href="tutorial.html" />
    <link rel="prev" title="Overview" href="overview.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="modules.html" class="icon icon-home"> MINORg
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Configuration</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#config-file">config file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#alias-lookups">Alias lookups</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#lookup-files-formats">Lookup files formats</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#cluster">cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="#genome">genome</a></li>
<li class="toctree-l4"><a class="reference internal" href="#reference">reference</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="minorg.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="modules.html">MINORg</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="modules.html" class="icon icon-home"></a> &raquo;</li>
      <li>Configuration</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/configuration.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="configuration">
<h1>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline"></a></h1>
<p>Note that all command line code snippets in the following tutorial are for <strong>bash terminal</strong>. You may have to adapt them according to your operating system.</p>
<section id="config-file">
<h2>config file<a class="headerlink" href="#config-file" title="Permalink to this headline"></a></h2>
<p>The config.ini file is used for two main purposes: setting default value and assigning short aliases to long values. All entries in the config.ini file that end with ‘alias’ or ‘sets’ are used to assign aliases, while the rest are for setting default values. Use the following to export the config file location as an environment variable:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">export</span> <span class="nv">MINORG_CONFIG</span><span class="o">=</span>/path/to/config.ini
</pre></div>
</div>
<p>If you’d like the config file to apply to all users, consider setting it globally.</p>
</section>
<section id="alias-lookups">
<h2>Alias lookups<a class="headerlink" href="#alias-lookups" title="Permalink to this headline"></a></h2>
<p>To fully understand how aliases are assigned, we shall use the sample_config.ini file (provided at <a class="reference external" href="https://github.com/rlrq/MINORg/blob/master/examples/sample_config.ini">https://github.com/rlrq/MINORg/blob/master/examples/sample_config.ini</a>) and zoom in on reference genomes as an example. First, let use look at <code class="docutils literal notranslate"><span class="pre">reference</span> <span class="pre">sets</span></code> under the section <code class="docutils literal notranslate"><span class="pre">[lookup]</span></code>. Here, we assign aliases for lookup files (which at the command line we can specify using <code class="docutils literal notranslate"><span class="pre">--reference-set</span></code>).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">reference</span> <span class="n">sets</span> <span class="o">=</span> <span class="n">athaliana</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">athaliana_genomes</span><span class="o">.</span><span class="n">txt</span>
                 <span class="n">arabidopsis</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">arabidopsis_genomes</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Each lookup file contains mapping of aliases to reference genome files as well as meta data. Let us look at the file athaliana_genomes.txt (provided at <a class="reference external" href="https://github.com/rlrq/MINORg/blob/master/examples/athaliana_genomes.txt">https://github.com/rlrq/MINORg/blob/master/examples/athaliana_genomes.txt</a>).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tair10</span><span class="p">;</span><span class="n">TAIR10</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_TAIR10</span><span class="o">.</span><span class="n">fasta</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_TAIR10</span><span class="o">.</span><span class="n">gff</span>  <span class="mi">1</span>
<span class="n">araport11</span>     <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araport11</span><span class="o">.</span><span class="n">fasta</span>     <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araport11</span><span class="o">.</span><span class="n">gff</span>       <span class="mi">1</span>
</pre></div>
</div>
<p>This is a tab-separated file where:</p>
<ul class="simple">
<li><p>column 1: semicolon-separated alias(es)</p></li>
<li><p>column 2: path to genome FASTA file</p></li>
<li><p>column 3: path to genome GFF3 file</p></li>
<li><p>column 4: NCBI genetic code number or name (optional if standard genetic code)</p></li>
<li><p>column 5: mapping of nonstandard GFF3 attribute field names to standard field names (optional if standard)</p></li>
</ul>
<p>By using <code class="docutils literal notranslate"><span class="pre">--reference-set</span> <span class="pre">athaliana</span></code> in the command line execution, we can use <code class="docutils literal notranslate"><span class="pre">--reference</span> <span class="pre">tair10</span></code> (or <code class="docutils literal notranslate"><span class="pre">--rference</span> <span class="pre">TAIR10</span></code>) to tell MINORg to use ‘/path/to/subset_ref_TAIR10.fasta’ and ‘/path/to/subset_ref_TAIR10.gff’ as reference assembly and annotation files respectively without having to type their paths out. In fact, we can even specify multiple reference genomes using <code class="docutils literal notranslate"><span class="pre">--reference</span> <span class="pre">tair10,araport11</span></code> and making sure to use a comma to separate reference genome aliases.</p>
<p>But there’s more! We can set ‘athaliana’ as the default reference set AND ‘tair10’ as the default reference in the <code class="docutils literal notranslate"><span class="pre">[data]</span></code> section of the config file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">reference</span> <span class="o">=</span> <span class="n">TAIR10</span>
<span class="n">reference</span> <span class="nb">set</span> <span class="o">=</span> <span class="n">athaliana</span>
</pre></div>
</div>
<p>This way, unless you wish to use a different reference genome from the default, you won’t have to type <code class="docutils literal notranslate"><span class="pre">--reference-set</span> <span class="pre">athaliana</span> <span class="pre">--reference</span> <span class="pre">tair10</span></code> either! This is particularly useful if you primarily design gRNA for only a single species. Of course, if you wish, you can combine all reference set files into a single massive file containing the mapping information for all possible reference genomes instead of having multiple files. However, I personally find it easier to maintain smaller files with descriptive names.</p>
<p>You can view the reference genomes available using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --references <span class="c1">## prints contents of default reference set file</span>
$ minorg --reference-set &lt;reference <span class="nb">set</span> alias&gt; --references <span class="c1">## prints contents of specified reference set file</span>
</pre></div>
</div>
<p>The same logic applies as well to <code class="docutils literal notranslate"><span class="pre">cluster</span> <span class="pre">sets</span></code>-<code class="docutils literal notranslate"><span class="pre">cluster</span> <span class="pre">set</span></code> and <code class="docutils literal notranslate"><span class="pre">genome</span> <span class="pre">sets</span></code>-<code class="docutils literal notranslate"><span class="pre">genome</span> <span class="pre">set</span></code>, with the caveat that there is no option to set default clusters or query genomes.</p>
<section id="lookup-files-formats">
<h3>Lookup files formats<a class="headerlink" href="#lookup-files-formats" title="Permalink to this headline"></a></h3>
<section id="cluster">
<h4>cluster<a class="headerlink" href="#cluster" title="Permalink to this headline"></a></h4>
<p>The lookup files specified in <code class="docutils literal notranslate"><span class="pre">cluster</span> <span class="pre">sets</span></code> should look like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">RPS4</span><span class="p">;</span><span class="n">TTR1</span><span class="p">;</span><span class="n">RPS4_TTR1</span>   <span class="n">AT5G45050</span><span class="p">,</span><span class="n">AT5G45060</span><span class="p">,</span><span class="n">AT5G45200</span><span class="p">,</span><span class="n">AT5G45210</span><span class="p">,</span><span class="n">AT5G45220</span><span class="p">,</span><span class="n">AT5G45230</span><span class="p">,</span><span class="n">AT5G45240</span><span class="p">,</span><span class="n">AT5G45250</span>
<span class="n">RPS6</span>  <span class="n">AT5G46260</span><span class="p">,</span><span class="n">AT5G46270</span><span class="p">,</span><span class="n">AT5G46450</span><span class="p">,</span><span class="n">AT5G46470</span><span class="p">,</span><span class="n">AT5G46490</span><span class="p">,</span><span class="n">AT5G46510</span><span class="p">,</span><span class="n">AT5G46520</span>
</pre></div>
</div>
<p>This is a tab-separated file where:</p>
<ul class="simple">
<li><p>column 1: semicolon-separated alias(es)</p></li>
<li><p>column 2: comma-separated gene IDs of genes in the cluster</p></li>
</ul>
</section>
<section id="genome">
<h4>genome<a class="headerlink" href="#genome" title="Permalink to this headline"></a></h4>
<p>The lookup files specified in <code class="docutils literal notranslate"><span class="pre">genome</span> <span class="pre">sets</span></code> should look like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">9654</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_9654</span><span class="o">.</span><span class="n">fasta</span>
<span class="mi">9655</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_9655</span><span class="o">.</span><span class="n">fasta</span>
<span class="mi">9944</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_9944</span><span class="o">.</span><span class="n">fasta</span>
<span class="mi">9947</span>  <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_9947</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<p>This is a tab-separated file where:</p>
<ul class="simple">
<li><p>column 1: semicolon-separated alias(es)</p></li>
<li><p>column 2: path to FASTA file</p></li>
</ul>
</section>
<section id="reference">
<h4>reference<a class="headerlink" href="#reference" title="Permalink to this headline"></a></h4>
<p>For ease of reference, the format for reference lookup files specified in <code class="docutils literal notranslate"><span class="pre">reference</span> <span class="pre">sets</span></code> is reproduced here:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tair10</span><span class="p">;</span><span class="n">TAIR10</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_TAIR10</span><span class="o">.</span><span class="n">fasta</span>        <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_TAIR10</span><span class="o">.</span><span class="n">gff</span>  <span class="mi">1</span>
<span class="n">araport11</span>     <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araport11</span><span class="o">.</span><span class="n">fasta</span>     <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">subset_ref_Araport11</span><span class="o">.</span><span class="n">gff</span>       <span class="mi">1</span>
</pre></div>
</div>
<p>This is a tab-separated file where:</p>
<ul class="simple">
<li><p>column 1: semicolon-separated alias(es)</p></li>
<li><p>column 2: path to genome FASTA file</p></li>
<li><p>column 3: path to genome GFF3 file</p></li>
<li><p>column 4: NCBI genetic code number or name (optional if standard genetic code)</p></li>
<li><p>column 5: mapping of nonstandard GFF3 attribute field names to standard field names (optional if standard)</p></li>
</ul>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="overview.html" class="btn btn-neutral float-left" title="Overview" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorial.html" class="btn btn-neutral float-right" title="Tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Rachelle R.Q. Lee.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>