<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Overview &mdash; MINORg 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Configuration" href="configuration.html" />
    <link rel="prev" title="Installation" href="install.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="modules.html" class="icon icon-home"> MINORg
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#command-line">Command line</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#subcommands">Subcommands</a></li>
<li class="toctree-l3"><a class="reference internal" href="#help-page">Help page</a></li>
<li class="toctree-l3"><a class="reference internal" href="#aliases">Aliases</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#python-package">Python package</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="minorg.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="modules.html">MINORg</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="modules.html" class="icon icon-home"></a> &raquo;</li>
      <li>Overview</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/overview.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this headline"></a></h1>
<p>(Note that all command line code snippets in the following tutorial are for <strong>bash terminal</strong>. You may have to adapt them according to your operating system.)</p>
<p>MINORg (<strong>MI</strong>nimum <strong>NO</strong>n-<strong>R</strong>eference <strong>g</strong>RNA) is a 4-part programme created to design a minimum number of gRNA to cover multiple non-reference targets. Nevertheless, MINORg is also capable of designing gRNA for one target as well as for reference genes. It is available as both a command line programme as well as a Python package.</p>
<p>The 4 broad steps of MINORg are as follows:</p>
<ul class="simple">
<li><p>seq: generation of target sequences</p></li>
<li><p>grna: generation of all potential gRNA from target sequences</p></li>
<li><p>filter: filtering of potential gRNA by GC, off-target, and/or within-feature</p></li>
<li><p>minimumset: generation of minimum set(s) of gRNA that cover all target sequences</p></li>
</ul>
<section id="command-line">
<h2>Command line<a class="headerlink" href="#command-line" title="Permalink to this headline"></a></h2>
<p>The command line interface is a wrapper for the Python package and is built using Typer (<a class="reference external" href="https://github.com/tiangolo/typer">https://github.com/tiangolo/typer</a>).</p>
<section id="subcommands">
<h3>Subcommands<a class="headerlink" href="#subcommands" title="Permalink to this headline"></a></h3>
<p>Each of the subcommands (<code class="docutils literal notranslate"><span class="pre">seq</span></code>, <code class="docutils literal notranslate"><span class="pre">grna</span></code>, <code class="docutils literal notranslate"><span class="pre">filter</span></code>, and <code class="docutils literal notranslate"><span class="pre">minimumset</span></code>) can be separately executed at the command line using:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg &lt;subcommmand&gt; &lt;arguments&gt;
</pre></div>
</div>
<p>If no subcommand is specified, MINORg will default to the full programme.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg &lt;arguments&gt;
</pre></div>
</div>
</section>
<section id="help-page">
<h3>Help page<a class="headerlink" href="#help-page" title="Permalink to this headline"></a></h3>
<p>To view the help page for each subcommand (which describes the subcommand and its parameters), use:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg &lt;subcommand&gt; --help
</pre></div>
</div>
<p>To view the help page for the full programme, use:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg full --help
</pre></div>
</div>
<p>Using the following will print a help page that lists common parameters and valid subcommand names:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --help
</pre></div>
</div>
</section>
<section id="aliases">
<h3>Aliases<a class="headerlink" href="#aliases" title="Permalink to this headline"></a></h3>
<p>Using a config.ini file, the command line version of MINORg allows users to supply short aliases in place of file names and/or combinations of parameters, as well as set default values for some parameters (such as the reference genome). For example, with the appropriate config.ini setup and lookup files, <code class="docutils literal notranslate"><span class="pre">--reference</span> <span class="pre">TAIR10</span></code> can be used in place of <code class="docutils literal notranslate"><span class="pre">--assembly</span> <span class="pre">/path/to/TAIR10/genome.fasta</span> <span class="pre">--annotation</span> <span class="pre">/path/to/TAIR10/genome.gff3</span></code>, and <code class="docutils literal notranslate"><span class="pre">--indv</span> <span class="pre">ler1</span></code> can be used in place of <code class="docutils literal notranslate"><span class="pre">--query</span> <span class="pre">/path/to/ler1.fasta</span></code>. For details on how to set up a config.ini file, see <a class="reference internal" href="configuration.html#configuration"><span class="std std-ref">Configuration</span></a> in the <a class="reference internal" href="tutorial.html#tutorial"><span class="std std-ref">Tutorial</span></a> section.</p>
</section>
</section>
<section id="python-package">
<h2>Python package<a class="headerlink" href="#python-package" title="Permalink to this headline"></a></h2>
<p>Unlike the command line version, the Python package does not support aliases or preset parameter combinations beyond optionally reading default values from a config.ini file. Nevertheless, most arguments at the command line have equivalents as attributes in the class <a class="reference internal" href="minorg.MINORg.html#minorg.MINORg.MINORg" title="minorg.MINORg.MINORg"><code class="xref py py-class docutils literal notranslate"><span class="pre">MINORg</span></code></a>. For a more comprehensive list of the similarities and differences, please refer to the <a class="reference internal" href="#parameters"><span class="std std-ref">Parameters</span></a> section below.</p>
</section>
<section id="parameters">
<h2>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline"></a></h2>
<p>Several CLI arguments have no equivalents in the Python module as they were intended to simplify the building of commands for users who have little to no experience with coding. Users of the Python package are assumed to be comfortable with generating their own preset parameter combinations.</p>
<p>The table below lists the major similarities and differences between CLI arguments and the Python package’s MINORg class attributes (note that some attriutes are in fact properties, but with the exception of <code class="docutils literal notranslate"><span class="pre">reference</span></code>, setting them should be no different from setting attributes).</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 18%" />
<col style="width: 26%" />
<col style="width: 26%" />
<col style="width: 30%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Category</p></td>
<td><p>CLI arguments</p></td>
<td><p>Python attributes</p></td>
<td><p>Description</p></td>
</tr>
<tr class="row-even"><td rowspan="3"><p>General</p></td>
<td><p>directory</p></td>
<td><p>directory</p></td>
<td><p>output directory</p></td>
</tr>
<tr class="row-odd"><td><p>prefix</p></td>
<td><p>prefix</p></td>
<td><p>output file/directory
prefix</p></td>
</tr>
<tr class="row-even"><td><p>thread</p></td>
<td><p>thread</p></td>
<td><p>threads</p></td>
</tr>
<tr class="row-odd"><td rowspan="3"><p>Executable</p></td>
<td><p>blastn</p></td>
<td><p>blastn</p></td>
<td><p>local BLAST’s blastn</p></td>
</tr>
<tr class="row-even"><td><p>rpsblast</p></td>
<td><p>rpsblast</p></td>
<td><p>local BLAST’s
rpsblast/rpsblast+</p></td>
</tr>
<tr class="row-odd"><td><p>mafft</p></td>
<td><p>mafft</p></td>
<td><p>MAFFT</p></td>
</tr>
<tr class="row-even"><td rowspan="7"><p>Reference
genomes</p>
<p>(CLI: seq,
full;</p>
<p>Python: seq,
filter)</p>
</td>
<td><p>reference &lt;alias&gt;</p></td>
<td><p>reference &lt;dict of
Reference object&gt;</p></td>
<td><p>reference genome</p></td>
</tr>
<tr class="row-odd"><td><p>assembly</p></td>
<td></td>
<td><p>reference genome FASTA</p></td>
</tr>
<tr class="row-even"><td><p>annotation</p></td>
<td></td>
<td><p>reference genome GFF</p></td>
</tr>
<tr class="row-odd"><td><p>attr_mod</p></td>
<td></td>
<td><p>mapping for non-standard
GFF attribute field names</p></td>
</tr>
<tr class="row-even"><td><p>genetic_code</p></td>
<td></td>
<td><p>NCBI genetic code number
or name</p></td>
</tr>
<tr class="row-odd"><td><p>ext_gene</p></td>
<td></td>
<td><p>FASTA file of genes to
add to reference genome</p></td>
</tr>
<tr class="row-even"><td><p>ext_cds</p></td>
<td></td>
<td><p>FASTA file of CDS of
genes to add to reference
genome</p></td>
</tr>
<tr class="row-odd"><td rowspan="8"><p>[seq]</p>
<p>target
definition</p>
</td>
<td><p>gene</p></td>
<td><p>gene<strong>s</strong></p></td>
<td><p>gene IDs</p></td>
</tr>
<tr class="row-even"><td><p>cluster</p></td>
<td></td>
<td><p>cluster aliases</p></td>
</tr>
<tr class="row-odd"><td><p>indv</p></td>
<td></td>
<td><p>individuals to discover
targets in</p></td>
</tr>
<tr class="row-even"><td><p>target</p></td>
<td><p>target</p></td>
<td><p>FASTA file of sequences
to find gRNA in</p></td>
</tr>
<tr class="row-odd"><td><p>query</p></td>
<td><p>query</p></td>
<td><p>FASTA file(s) to discover
targets in</p></td>
</tr>
<tr class="row-even"><td><p>domain &lt;alias&gt;</p></td>
<td></td>
<td><p>aliases of domains to
find gRNA in</p></td>
</tr>
<tr class="row-odd"><td><p>domain &lt;Pssm-Id&gt;</p></td>
<td><p>pssm_ids</p></td>
<td><p>Pssm-Id(s) of domains to
find gRNA in</p></td>
</tr>
<tr class="row-even"><td></td>
<td><p>domain_name</p></td>
<td><p>human-readable domain
name used in sequence and
file names in place of
Pssm-Ids</p></td>
</tr>
<tr class="row-odd"><td rowspan="7"><p>[seq]</p>
<p>inferring
homologues</p>
<p>from BLASTN
hits</p>
</td>
<td><p>minid</p></td>
<td><p>minid</p></td>
<td><p>minimum hit % identity</p></td>
</tr>
<tr class="row-even"><td><p>minlen</p></td>
<td><p>minlen</p></td>
<td><p>minimum merged hits
length</p></td>
</tr>
<tr class="row-odd"><td><p>mincdslen</p></td>
<td><p>mincdslen</p></td>
<td><p>minimum CDS length of
merged hits</p></td>
</tr>
<tr class="row-even"><td><p>check_recip</p></td>
<td><p>check_recip</p></td>
<td><p>execute reciprocal check</p></td>
</tr>
<tr class="row-odd"><td><p>relax_recip</p></td>
<td><p>relax_recip</p></td>
<td><p>execute relaxed
reciprocal check</p></td>
</tr>
<tr class="row-even"><td><p>merge_within</p></td>
<td><p>merge_within</p></td>
<td><p>maximum distance between
hits for merging</p></td>
</tr>
<tr class="row-odd"><td><p>check_id_before_merge</p></td>
<td><p>check_id_before_merge</p></td>
<td><p>filter hits by % identity
before merging</p></td>
</tr>
<tr class="row-even"><td rowspan="2"><p>[seq]</p>
<p>RPS-BLAST
options</p>
</td>
<td><p>db</p></td>
<td><p>db</p></td>
<td><p>path to local RPS-BLAST
database</p></td>
</tr>
<tr class="row-odd"><td><p>remote_rps</p></td>
<td><p>remote_rps</p></td>
<td><p>use remote RPS-BLAST
database (currently
non-functional)</p></td>
</tr>
<tr class="row-even"><td rowspan="2"><p>[grna]</p></td>
<td><p>pam</p></td>
<td><p>pam</p></td>
<td><p>PAM pattern</p></td>
</tr>
<tr class="row-odd"><td><p>length</p></td>
<td><p>length</p></td>
<td><p>gRNA length</p></td>
</tr>
<tr class="row-even"><td rowspan="2"><p>[filter]</p>
<p>GC</p>
</td>
<td><p>gc_min</p></td>
<td><p>gc_min</p></td>
<td><p>minimum GC content</p></td>
</tr>
<tr class="row-odd"><td><p>gc_max</p></td>
<td><p>gc_max</p></td>
<td><p>maximum GC content</p></td>
</tr>
<tr class="row-even"><td rowspan="4"><p>[filter]</p>
<p>feature</p>
</td>
<td><p>feature</p></td>
<td><p>feature</p></td>
<td><p>GFF3 feature type</p></td>
</tr>
<tr class="row-odd"><td><p>max_insertion</p></td>
<td><p>max_insertion</p></td>
<td><p>maximum allowable
insertion in feature</p></td>
</tr>
<tr class="row-even"><td><p>min_within_n</p></td>
<td><p>min_within_n</p></td>
<td><p>minimum number of
reference genes which
features overlap with
gRNA range in alignment</p></td>
</tr>
<tr class="row-odd"><td><p>min_within_fraction</p></td>
<td><p>min_within_fraction</p></td>
<td><p>minimum fraction of
reference genes which
features overlap with
gRNA range in alignment</p></td>
</tr>
<tr class="row-even"><td rowspan="12"><p>[filter]</p>
<p>background</p>
</td>
<td><p>background</p></td>
<td><p>background</p></td>
<td><p>FASTA files in which to
search for potential
off-targets</p></td>
</tr>
<tr class="row-odd"><td><p>screen_reference</p></td>
<td><p>screen_reference</p></td>
<td><p>include reference genomes
in search for potential
off-targets</p></td>
</tr>
<tr class="row-even"><td></td>
<td><p>mask</p></td>
<td><p>FASTA files of additional
sequences to mask</p></td>
</tr>
<tr class="row-odd"><td><p>unmask_ref</p></td>
<td></td>
<td><p>unmask reference genes</p></td>
</tr>
<tr class="row-even"><td><p>mask_gene</p></td>
<td></td>
<td><p>additional genes to mask</p></td>
</tr>
<tr class="row-odd"><td><p>unmask_gene</p></td>
<td></td>
<td><p>genes to unmask</p></td>
</tr>
<tr class="row-even"><td><p>mask_cluster</p></td>
<td></td>
<td><p>additional clusters to
mask</p></td>
</tr>
<tr class="row-odd"><td><p>unmask_cluster</p></td>
<td></td>
<td><p>clusters to unmask</p></td>
</tr>
<tr class="row-even"><td><p>ot_pamless</p></td>
<td><p>ot_pamless</p></td>
<td><p>ignore absense of PAM for
potential off-targets</p></td>
</tr>
<tr class="row-odd"><td><p>ot_mismatch</p></td>
<td><p>ot_mismatch</p></td>
<td><p>minimum acceptable
mismatches for
off-targets</p></td>
</tr>
<tr class="row-even"><td><p>ot_gap</p></td>
<td><p>ot_gap</p></td>
<td><p>minimum acceptable gaps
for off-targets</p></td>
</tr>
<tr class="row-odd"><td><p>skip_bg_check</p></td>
<td></td>
<td><p>skip off-target check</p></td>
</tr>
<tr class="row-even"><td><p>[filter]
exclude</p></td>
<td><p>exclude</p></td>
<td><p>exclude</p></td>
<td><p>FASTA file of gRNA
sequences to exclude</p></td>
</tr>
<tr class="row-odd"><td rowspan="5"><p>[minimumset]</p></td>
<td><p>accept_invalid</p></td>
<td><p>accept_invalid</p></td>
<td><p>score ‘NA’ as ‘pass’</p></td>
</tr>
<tr class="row-even"><td><p>accept_feature_unknown</p></td>
<td><p>accept_feature_unknown</p></td>
<td><p>score ‘NA’ as ‘pass’ for
feature check</p></td>
</tr>
<tr class="row-odd"><td></td>
<td><p>accept_invalid_field</p></td>
<td><p>score ‘NA’ as ‘pass’ if
all entries for a check
are ‘NA’</p></td>
</tr>
<tr class="row-even"><td><p>sets</p></td>
<td><p>sets</p></td>
<td><p>number of gRNA sets</p></td>
</tr>
<tr class="row-odd"><td><p>auto</p></td>
<td><p>auto</p></td>
<td><p>generate sets without
require manual user
confirmation for each set</p></td>
</tr>
</tbody>
</table>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="install.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="configuration.html" class="btn btn-neutral float-right" title="Configuration" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Rachelle R.Q. Lee.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>